autora/controller/protocol.py

```diff
@@ -0,0 +1,107 @@
+from enum import Enum
+from typing import Any, Dict, Optional, Protocol, Sequence, Set, TypeVar, Union
+
+from numpy.typing import ArrayLike
+from sklearn.base import BaseEstimator
+
+from autora.variable import VariableCollection
+
+State = TypeVar("State")
+
+
+class ResultKind(str, Enum):
+    """
+    Kinds of results which can be held in the Result object.
+
+    Examples:
+        >>> ResultKind.CONDITION is ResultKind.CONDITION
+        True
+
+        >>> ResultKind.CONDITION is ResultKind.METADATA
+        False
+
+        >>> ResultKind.CONDITION == "CONDITION"
+        True
+
+        >>> ResultKind.CONDITION == "METADATA"
+        False
+
+        >>> ResultKind.CONDITION in {ResultKind.CONDITION, ResultKind.PARAMS}
+        True
+
+        >>> ResultKind.METADATA in {ResultKind.CONDITION, ResultKind.PARAMS}
+        False
+    """
+
+    CONDITION = "CONDITION"
+    OBSERVATION = "OBSERVATION"
+    THEORY = "THEORY"
+    PARAMS = "PARAMS"
+    METADATA = "METADATA"
+
+    def __repr__(self):
+        cls_name = self.__class__.__name__
+        return f"{cls_name}.{self.name}"
+
+
+class SupportsDataKind(Protocol):
+    """Object with attributes for `data` and `kind`"""
+
+    data: Optional[Any]
+    kind: Optional[ResultKind]
+
+
+class SupportsControllerStateFields(Protocol):
+    """Support representing snapshots of a controller state as mutable fields."""
+
+    metadata: VariableCollection
+    params: Dict
+    conditions: Sequence[ArrayLike]
+    observations: Sequence[ArrayLike]
+    theories: Sequence[BaseEstimator]
+
+    def update(self: State, **kwargs) -> State:
+        ...
+
+
+class SupportsControllerStateProperties(Protocol):
+    """Support representing snapshots of a controller state as immutable properties."""
+
+    def update(self: State, **kwargs) -> State:
+        ...
+
+    @property
+    def metadata(self) -> VariableCollection:
+        ...
+
+    @property
+    def params(self) -> Dict:
+        ...
+
+    @property
+    def conditions(self) -> Sequence[ArrayLike]:
+        ...
+
+    @property
+    def observations(self) -> Sequence[ArrayLike]:
+        ...
+
+    @property
+    def theories(self) -> Sequence[BaseEstimator]:
+        ...
+
+
+SupportsControllerState = Union[
```

Comment:
The `Supports` classes are 1) used for typing, and 2) we combine two because we have two valid but type-incompatible variants.

----

**Part 1:** These `Supports` classes are used for typing.

If you've got something like this:
```python
class SupportsParamsDictField(Protocol):
    """ Object which has a `params` attribute"""
    params: Dict[name, Any]

def print_params_from_cycle(cycle: SupportsParamsDictField):
    for key, value in cycle.params:
        print(f"{key}: {value}")
```

... then you're saying "I have a function which depends on the fact that I can access the params field of the object I pass it. Everything else I don't care about, so I don't need to specify that it is a particular _class_ of object."

The benefit is that you can then implement a bunch of different objects with the same interface, but that you don't have to care about whether they inherit from each other. If on the other hand you do this:
```python
def print_params_from_cycle(cycle: SimpleCycle):
    for key, value in cycle.params:
        print(f"{key}: {value}")
```
... then the type checker will complain if you give it anything except an object which is or inherits from the `SimpleCycle`. So then you've got to add the classes to the signature:

```python
def print_params_from_cycle(cycle: Union[SimpleCycle, BaseController, Pipeline]):
    for key, value in cycle.params:
        print(f"{key}: {value}")
```

... to let it pass. 

However, as soon as the user makes their own controller which doesn't inherit from our BaseController class, they can't just go in and update the signature (it's in the package which they can't easily modify) and their type checker will complain. They will turn the type checker off and then will lose those benefits.

----

**Part 2**: we have two equivalently valid but type-incompatible variants.

Our `History` and `Snapshot` objects have the same `params` interface from the user perspective – you can do:
```python
history.params == {<the params>}  # TRUE
snapshot.params == {<the params>}  # TRUE
```
and it's fine.

However, under the hood, the Snapshot uses a data class field, whereas the History uses a property, and fulfills the following Protocol instead:

```python
class SupportsParamsDictProperty(Protocol):
    """ Object which has a `params` property"""
    @property
    def params(self) -> Dict[name, Any]:
        ...
```

If you try to do something like this:
```python
history = History(<data>)  # SupportsParamsDictProperty
print_params_from_cycle(history)
```
... you'll get a typing error, because the `@property` decorator makes the `SupportsParamsDictProperty.params` attribute read-only, whereas the `SupportsParamsDictField.params` field is read-write. The history doesn't fulfill the required protocol, which is read/write.

So, you've got to say "my function can work with both of these variants quite happily":
```python
SupportsParamsDict = Union[SupportsParamsDictField, SupportsParamsDictProperty]

def print_params_from_cycle(cycle: SupportsParamsDict):
    ... <as above>
```

The `SupportsControllerState` does exactly that, just with more fields.